File: backend/src/main/java/com/yourspace/article/service/ArticleService.java
package com.yourspace.article.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.yourspace.article.dto.ArticleResponse;
import com.yourspace.article.dto.ArticleSummaryResponse;
import com.yourspace.article.entity.ArticleStatus;
import com.yourspace.common.PageResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
public class ArticleService {
    private static final Logger log = LoggerFactory.getLogger(ArticleService.class);
    private static final String METADATA_FILE = "article.json";
    private static final String CONTENT_FILE = "article.md";

    private final ObjectMapper objectMapper;
    private final Path articlesRoot;

    public ArticleService(ObjectMapper objectMapper, @Value("${app.articles.directory:./public/articles}") String articlesDirectory) {
        this.objectMapper = objectMapper;
        this.articlesRoot = Paths.get(articlesDirectory).toAbsolutePath().normalize();
    }

    public PageResponse<ArticleSummaryResponse> listPublished(int page, int size, String tag) {
        String normalizedTag = normalizeTag(tag);
        List<ArticleMetadata> articles = readMetadata().stream()
                .filter(article -> article.status() == ArticleStatus.PUBLISHED)
                .filter(article -> normalizedTag == null || article.tags() != null && article.tags().stream().map(this::normalizeTag).anyMatch(normalizedTag::equals))
                .sorted(Comparator.comparing(ArticleMetadata::publishedAt, Comparator.nullsLast(Comparator.reverseOrder()))
                        .thenComparing(ArticleMetadata::createdAt, Comparator.nullsLast(Comparator.reverseOrder())))
                .toList();

        int from = Math.min(page * size, articles.size());
        int to = Math.min(from + size, articles.size());
        int totalPages = articles.isEmpty() ? 0 : (int) Math.ceil(articles.size() / (double) size);
        List<ArticleSummaryResponse> content = articles.subList(from, to).stream()
                .map(this::toSummary)
                .toList();
        return new PageResponse<>(content, page, size, articles.size(), totalPages);
    }

    public ArticleResponse getPublished(String slug) {
        ArticleMetadata metadata = readMetadata().stream()
                .filter(article -> article.status() == ArticleStatus.PUBLISHED)
                .filter(article -> normalizeSlug(article.slug()).equals(normalizeSlug(slug)))
                .findFirst()
                .orElseThrow(() -> new ArticleNotFoundException(slug));

        Path folder = folderFor(metadata);
        Path contentPath = folder.resolve(CONTENT_FILE).normalize();
        if (!contentPath.startsWith(folder) || !Files.isRegularFile(contentPath)) {
            throw new ArticleNotFoundException(slug);
        }
        try {
            String markdown = Files.readString(contentPath, StandardCharsets.UTF_8);
            return new ArticleResponse(metadata.id(), metadata.slug(), metadata.title(), metadata.summary(), coverUrl(metadata, folder), markdown, tags(metadata), metadata.status(), metadata.publishedAt(), metadata.createdAt(), metadata.updatedAt());
        } catch (IOException exception) {
            log.warn("Could not read article content for {}", slug, exception);
            throw new ArticleNotFoundException(slug);
        }
    }

    private List<ArticleMetadata> readMetadata() {
        if (!Files.isDirectory(articlesRoot)) return List.of();
        List<ArticleMetadata> result = new ArrayList<>();
        try (Stream<Path> folders = Files.list(articlesRoot)) {
            folders.filter(Files::isDirectory).forEach(folder -> {
                Path metadataPath = folder.resolve(METADATA_FILE);
                if (!Files.isRegularFile(metadataPath)) return;
                try {
                    ArticleMetadata metadata = objectMapper.readValue(metadataPath.toFile(), ArticleMetadata.class);
                    if (metadata.slug() == null || !normalizeSlug(metadata.slug()).equals(normalizeSlug(folder.getFileName().toString()))) {
                        log.warn("Skipping article with mismatched folder and slug: {}", metadataPath);
                        return;
                    }
                    result.add(metadata);
                } catch (IOException | RuntimeException exception) {
                    log.warn("Skipping invalid article metadata: {}", metadataPath, exception);
                }
            });
        } catch (IOException exception) {
            log.warn("Could not scan article directory {}", articlesRoot, exception);
        }
        return result;
    }

    private ArticleSummaryResponse toSummary(ArticleMetadata metadata) {
        return new ArticleSummaryResponse(metadata.id(), metadata.slug(), metadata.title(), metadata.summary(), coverUrl(metadata, folderFor(metadata)), tags(metadata), metadata.status(), metadata.publishedAt(), metadata.createdAt(), metadata.updatedAt());
    }

    private Path folderFor(ArticleMetadata metadata) {
        return articlesRoot.resolve(normalizeSlug(metadata.slug())).normalize();
    }

    private String coverUrl(ArticleMetadata metadata, Path folder) {
        if (metadata.cover() == null || metadata.cover().isBlank()) return null;
        Path coverPath = folder.resolve(metadata.cover()).normalize();
        if (!coverPath.startsWith(folder) || !Files.isRegularFile(coverPath)) return null;
        return "/articles/" + normalizeSlug(metadata.slug()) + "/" + metadata.cover().replace('\\', '/');
    }

    private Set<String> tags(ArticleMetadata metadata) {
        if (metadata.tags() == null) return Set.of();
        return metadata.tags().stream().filter(tag -> tag != null && !tag.isBlank()).map(this::normalizeTag).collect(Collectors.toCollection(LinkedHashSet::new));
    }

    private String normalizeSlug(String slug) {
        return slug == null ? "" : slug.trim().toLowerCase(Locale.ROOT);
    }

    private String normalizeTag(String tag) {
        return tag == null || tag.isBlank() ? null : tag.trim().toLowerCase(Locale.ROOT);
    }
}

File: backend-python/app/article_service.py
"""Read published articles from one-folder-per-article content storage."""

from __future__ import annotations

import json
import logging
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .models import ArticleDetail, ArticlePageResponse, ArticleSummary
from .markdown_io import read_markdown

logger = logging.getLogger(__name__)


class ArticleNotFoundError(Exception):
    def __init__(self, slug: str) -> None:
        super().__init__(f"Article not found: {slug}")
        self.slug = slug


@dataclass(frozen=True)
class ArticleMetadata:
    id: str | None
    slug: str
    title: str
    summary: str | None
    tags: tuple[str, ...]
    status: str
    published_at: str | None
    created_at: str | None
    updated_at: str | None
    cover: str | None
    published_sort: datetime | None
    created_sort: datetime | None


def _normalize_slug(value: str | None) -> str:
    return (value or "").strip().lower()


def _normalize_tag(value: str | None) -> str | None:
    if value is None or not value.strip():
        return None
    return value.strip().lower()


def _parse_timestamp(value: Any, field: str, path: Path) -> tuple[str | None, datetime | None]:
    if value is None or value == "":
        return None, None
    if not isinstance(value, str):
        raise ValueError(f"{field} must be an ISO timestamp")
    normalized = value.replace("Z", "+00:00")
    parsed = datetime.fromisoformat(normalized)
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return value, parsed


class ArticleService:
    def __init__(self, articles_root: Path) -> None:
        self.articles_root = articles_root

    def list_published(self, page: int, size: int, tag: str | None, q: str | None = None) -> ArticlePageResponse:
        normalized_tag = _normalize_tag(tag)
        normalized_query = q.strip().lower() if q and q.strip() else None
        articles = [
            article
            for article in self._read_metadata()
            if article.status == "PUBLISHED"
            and (
                normalized_tag is None
                or normalized_tag in {_normalize_tag(article_tag) for article_tag in article.tags}
            )
            and (
                normalized_query is None
                or normalized_query in "\n".join((article.title, article.summary or "", " ".join(article.tags), self._read_body_for_query(article))).lower()
            )
        ]
        articles.sort(
            key=lambda article: (
                article.published_sort is not None,
                article.published_sort or datetime.min.replace(tzinfo=timezone.utc),
                article.created_sort is not None,
                article.created_sort or datetime.min.replace(tzinfo=timezone.utc),
            ),
            reverse=True,
        )

        start = min(page * size, len(articles))
        end = min(start + size, len(articles))
        total_pages = 0 if not articles else (len(articles) + size - 1) // size
        return ArticlePageResponse(
            content=[self._to_summary(article) for article in articles[start:end]],
            page=page,
            size=size,
            totalElements=len(articles),
            totalPages=total_pages,
        )

    def get_published(self, slug: str) -> ArticleDetail:
        metadata = next(
            (
                article
                for article in self._read_metadata()
                if article.status == "PUBLISHED"
                and _normalize_slug(article.slug) == _normalize_slug(slug)
            ),
            None,
        )
        if metadata is None:
            raise ArticleNotFoundError(slug)

        folder = self._folder_for(metadata)
        content_path = (folder / "article.md").resolve()
        if not self._is_inside(content_path, folder) or not content_path.is_file():
            raise ArticleNotFoundError(slug)
        try:
            markdown = read_markdown(content_path)
        except OSError as exc:
            logger.warning("Could not read article content for %s: %s", slug, exc)
            raise ArticleNotFoundError(slug) from exc

        summary = self._to_summary(metadata)
        return ArticleDetail(**summary.model_dump(), contentMarkdown=markdown)

    def _read_body_for_query(self, metadata: ArticleMetadata) -> str:
        if not metadata.slug:
            return ""
        path = (self.articles_root / _normalize_slug(metadata.slug) / "article.md").resolve()
        folder = (self.articles_root / _normalize_slug(metadata.slug)).resolve()
        if not self._is_inside(path, folder) or not path.is_file():
            return ""
        try:
            return read_markdown(path)
        except OSError:
            return ""

    def _read_metadata(self) -> list[ArticleMetadata]:
        if not self.articles_root.is_dir():
            return []
        result: list[ArticleMetadata] = []
        try:
            folders = list(self.articles_root.iterdir())
        except OSError as exc:
            logger.warning("Could not scan article directory %s: %s", self.articles_root, exc)
            return []

        for folder in folders:
            if not folder.is_dir():
                continue
            metadata_path = folder / "article.json"
            if not metadata_path.is_file():
                continue
            try:
                raw = json.loads(metadata_path.read_text(encoding="utf-8"))
                metadata = self._parse_metadata(raw, metadata_path)
                if _normalize_slug(metadata.slug) != _normalize_slug(folder.name):
                    logger.warning("Skipping article with mismatched folder and slug: %s", metadata_path)
                    continue
                result.append(metadata)
            except (OSError, ValueError, TypeError, json.JSONDecodeError) as exc:
                logger.warning("Skipping invalid article metadata %s: %s", metadata_path, exc)
        return result

    def _parse_metadata(self, raw: Any, path: Path) -> ArticleMetadata:
        if not isinstance(raw, dict):
            raise ValueError("metadata must be an object")
        slug = raw.get("slug")
        title = raw.get("title")
        if not isinstance(slug, str) or not slug.strip():
            raise ValueError("slug is required")
        if not isinstance(title, str) or not title.strip():
            raise ValueError("title is required")

        published_at, published_sort = _parse_timestamp(raw.get("publishedAt"), "publishedAt", path)
        created_at, created_sort = _parse_timestamp(raw.get("createdAt"), "createdAt", path)
        updated_at, _ = _parse_timestamp(raw.get("updatedAt"), "updatedAt", path)
        raw_tags = raw.get("tags") or []
        if not isinstance(raw_tags, list):
            raise ValueError("tags must be an array")
        tags = tuple(
            tag.strip()
            for tag in raw_tags
            if isinstance(tag, str) and tag.strip()
        )
        return ArticleMetadata(
            id=str(raw["id"]) if raw.get("id") is not None else None,
            slug=slug.strip(),
            title=title,
            summary=raw.get("summary") if isinstance(raw.get("summary"), str) else None,
            tags=tags,
            status=str(raw.get("status") or "").upper(),
            published_at=published_at,
            created_at=created_at,
            updated_at=updated_at,
            cover=raw.get("cover") if isinstance(raw.get("cover"), str) else None,
            published_sort=published_sort,
            created_sort=created_sort,
        )

    def _to_summary(self, metadata: ArticleMetadata) -> ArticleSummary:
        return ArticleSummary(
            id=metadata.id,
            slug=metadata.slug,
            title=metadata.title,
            summary=metadata.summary,
            coverUrl=self._cover_url(metadata),
            tags=list(dict.fromkeys(
                normalized for normalized in (_normalize_tag(tag) for tag in metadata.tags)
                if normalized is not None
            )),
            status=metadata.status,
            publishedAt=metadata.published_at,
            createdAt=metadata.created_at,
            updatedAt=metadata.updated_at,
        )

    def _folder_for(self, metadata: ArticleMetadata) -> Path:
        return (self.articles_root / _normalize_slug(metadata.slug)).resolve()

    def _cover_url(self, metadata: ArticleMetadata) -> str | None:
        if not metadata.cover or not metadata.cover.strip():
            return None
        folder = self._folder_for(metadata)
        cover_path = (folder / metadata.cover).resolve()
        if not self._is_inside(cover_path, folder) or not cover_path.is_file():
            return None
        relative_cover = metadata.cover.replace("\\", "/").lstrip("/")
        return f"/articles/{_normalize_slug(metadata.slug)}/{relative_cover}"

    @staticmethod
    def _is_inside(path: Path, parent: Path) -> bool:
        try:
            path.relative_to(parent)
            return True
        except ValueError:
            return False
